components/server/src/oauth-server/oauth-controller.ts

```diff
@@ -0,0 +1,148 @@
+
+                // Have they authorized the local-app?
+                const wasApproved = req.query['approved'] || '';
+                log.info(`APPROVED?: ${wasApproved}`)
+                if (wasApproved === 'no') {
+                    // Let the local app know they rejected the approval
+                    const rt = req.query.redirect_uri;
+                    if (!rt || !rt.startsWith("http://localhost:")) {
+                        log.error(`/oauth/authorize: invalid returnTo URL: "${rt}"`)
+                        res.sendStatus(400);
+                        return;
+                    }
+                    res.redirect(`${rt}/?approved=no`);
+                    return;
+                }
+
+                const oauthClientsApproved = user?.additionalData?.oauthClientsApproved;
+                const clientID = localAppClientID;
+                if (!oauthClientsApproved || !oauthClientsApproved[clientID]) {
+                    const client = await authorizationServer.getClientByIdentifier(clientID)
+                    if (client) {
+                        const redirectTarget = encodeURIComponent(`${this.env.hostUrl}api${req.originalUrl}`);
+                        const redirectTo = `${this.env.hostUrl}oauth-approval?clientID=${client.id}&clientName=${client.name}&returnTo=${redirectTarget}`;
+                        log.info(`AUTH Redirecting to approval: ${redirectTo}`);
+                        res.redirect(redirectTo)
+                        return;
+                    } else {
+                        log.error(`/oauth/authorize unknown client id: "${clientID}"`)
+                        res.sendStatus(400);
+                        return;
+                    }
+                }
+
+                const request = new OAuthRequest(req);
+
+                try {
+                    // Validate the HTTP request and return an AuthorizationRequest object.
+                    const authRequest = await authorizationServer.validateAuthorizationRequest(request);
+
+                    // Once the user has logged in set the user on the AuthorizationRequest
+                    authRequest.user = { id: user.id }
+                    console.log(`user has logged in - setting the user on the AuthorizationRequest ${JSON.stringify(user)}, ${JSON.stringify(authRequest)}`);
+
+                    // The user has approved the client so update the status
+                    authRequest.isAuthorizationApproved = true;
+
+                    // Return the HTTP redirect response
+                    const oauthResponse = await authorizationServer.completeAuthorizationRequest(authRequest);
+                    return handleResponse(req, res, oauthResponse);
+                } catch (e) {
+                    handleError(e, res);
+                }
+            });
+
+            router.post("/oauth/token", async (req: express.Request, res: express.Response) => {
+                log.info(`TOKEN: ${JSON.stringify(req.body)}`);
```

Comment:
nit: it's the code at this point

Reply:
Not sure I understand the comment, can you elaborate pls?